File: src/deig/fitness.clj
(ns deig.fitness
    (:require
      [libpython-clj.require :refer [require-python]]
      [libpython-clj.python :as py :refer [call-attr]]
      [quil.core :as q]))

(require-python
  '[numpy :as np]
  '[keras_preprocessing.image :as image-processing]
  '[tensorflow.keras.models :as models])

;; Load default Keras MNIST classifier
(def model (models/load_model "mnist.h5"))

(defn fitness [img]
  "Evaluate an image based on its best categorical score"
    (let [x-array (image-processing/img_to_array img)
          x (np/expand_dims (np/array x-array) 0)
          probabilities (first (call-attr model "predict_proba" x))
          fitness (* (apply max probabilities) 100)]
         fitness))

(defn predict-class [img]
 "Evaluate an image based on its best categorical score"
   (let [x-array (image-processing/img_to_array img)
         x (np/expand_dims (np/array x-array) 0)
         class (first (call-attr model "predict_classes" x))]
        class))

(defn save-image [genome generation class]
  "Save individual's genome to file"
  (call-attr (image-processing/array_to_img genome) "save" (str "./individuals/gen" generation "predicted" class ".png")))

;; Example code
(defn grayscale-genome [dimension]
  "Returns random grayscale genome of dimensions dimension*dimension"
  (vec (repeatedly dimension
                   #(vec (repeatedly dimension (fn [] (vector (* (rand) 256))))))))

(defn line-genome []
  "Returns a genome with a centered vertical white line."
  (let [genome (grayscale-genome 28)]
    (vec (map-indexed (fn [i col]
                        (map-indexed (fn [j pixel]
                                       (if (and (< j 16) (> j 13) (> i 4) (< i 26))
                                         [255]
                                         [0])) col)) genome))))

(defn trial
    "Run a practice fitness run"
    []
  (fitness (line-genome)))


;: Should return 100.
#_ (trial)     

File: fitness.py
'''
This code is the python version of fitness.clj
'''

# Utility imports for the example
from discriminator import Discriminator
import os
import random

# Imports our Clojure code will have to use
import numpy as np

from keras_preprocessing import image
from tensorflow.keras.models import load_model


def fitness(img):
    '''
    Our fitness function to be copied to Clojure
    '''
    x = image.img_to_array(img)
    x = np.expand_dims(x, axis=0)

    model = load_model('./mnist.h5')

    # This prediction variable is not used in fitness calculation
    prediction = model.predict_classes(x)
    print('Prediction is ', prediction)

    probabilities = model.predict_proba(x, verbose=1)[0]
    print('Probabilities are', probabilities)

    predictions = model.predict(x)
    print('New prediction is', predictions)

    # Fitness is calculated 0-100, where 100 is perfect and 0 is very bad
    fitness = max(probabilities) * 100
    print('Fitness is', fitness)
    return fitness

if __name__ == '__main__':
    'Sample usage'

    # Load file from HD
    img = image.load_img('test.jpg',
                        target_size=(28, 28), color_mode="grayscale")

    # Pass image to our fitness function
    fitness(img)
